backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/ColumnarCachedBatchSerializer.scala

```diff
@@ -0,0 +1,311 @@
+      conf: SQLConf): RDD[CachedBatch] = {
+    val localSchema = toStructType(schema)
+    if (!validateSchema(localSchema)) {
+      // we can not use columnar cache here, as the `RowToColumnar` does not support this schema
+      return rowBasedCachedBatchSerializer.convertInternalRowToCachedBatch(
+        input,
+        schema,
+        storageLevel,
+        conf)
+    }
+
+    // note, these metrics are unused but just make `RowToVeloxColumnarExec` happy
+    val metrics = BackendsApiManager.getMetricsApiInstance.genRowToColumnarMetrics(
+      SparkSession.getActiveSession.orNull.sparkContext)
+    val numInputRows = metrics("numInputRows")
+    val numOutputBatches = metrics("numOutputBatches")
+    val convertTime = metrics("convertTime")
+    val numRows = conf.columnBatchSize
+    val rddColumnarBatch = input.mapPartitions {
+      it =>
+        RowToVeloxColumnarExec.toColumnarBatchIterator(
+          it,
+          localSchema,
+          numInputRows,
+          numOutputBatches,
+          convertTime,
+          numRows
+        )
+    }
+    convertColumnarBatchToCachedBatch(rddColumnarBatch, schema, storageLevel, conf)
+  }
+
+  override def convertCachedBatchToInternalRow(
+      input: RDD[CachedBatch],
+      cacheAttributes: Seq[Attribute],
+      selectedAttributes: Seq[Attribute],
+      conf: SQLConf): RDD[InternalRow] = {
+    if (!validateSchema(cacheAttributes)) {
+      // if we do not support this schema that means we are using row-based serializer,
+      // see `convertInternalRowToCachedBatch`, so fallback to vanilla Spark serializer
+      return rowBasedCachedBatchSerializer.convertCachedBatchToInternalRow(
+        input,
+        cacheAttributes,
+        selectedAttributes,
+        conf)
+    }
+
+    // note, these metrics are unused but just make `VeloxColumnarToRowExec` happy
+    val metrics = BackendsApiManager.getMetricsApiInstance.genColumnarToRowMetrics(
+      SparkSession.getActiveSession.orNull.sparkContext)
+    val numOutputRows = metrics("numOutputRows")
+    val numInputBatches = metrics("numInputBatches")
+    val convertTime = metrics("convertTime")
+    val rddColumnarBatch =
+      convertCachedBatchToColumnarBatch(input, cacheAttributes, selectedAttributes, conf)
+    rddColumnarBatch.mapPartitions {
+      it =>
+        VeloxColumnarToRowExec.toRowIterator(
+          it,
+          selectedAttributes,
+          numOutputRows,
+          numInputBatches,
+          convertTime
+        )
+    }
+  }
+
+  override def convertColumnarBatchToCachedBatch(
+      input: RDD[ColumnarBatch],
+      schema: Seq[Attribute],
+      storageLevel: StorageLevel,
+      conf: SQLConf): RDD[CachedBatch] = {
+    val timezoneId = SQLConf.get.sessionLocalTimeZone
+    val localSchema = toStructType(schema)
+    input.mapPartitions {
+      it =>
+        val executionCtxHandle = ExecutionCtxs.contextInstance().getHandle
+        val allocator = ArrowBufferAllocators.contextInstance()
```

Comment:
It will use ArrowBufferAllocator to allocate CachedBatch memory, so it will use offheap memory managed by gluten memory manager when doing df.cache. Is my understanding correct?

Reply:
No, the arrow buf is an intermediate thing, we finally will store array[byte] as rdd block cache which is onheap.